### ISSUE TITLE
Recommended asynchronous usage pattern of SerialPort

### ISSUE THREAD

**Issue by LeonidVasilyev** [dotnet/corefx#36040]
The question is about common command scenario when you send a command to a device and want to check a response to see the result or simply to check that the command is acknowledged. What is the recommended way to organize that in a non blocking fashion in general and in an UI application? 

Currently I see three options. First is `SerialPort.BaseStream.*Async` methods. `SerialPort.BaseStream.ReadAsync()` uses `Stream.ReadAsync()` implementation and ignores `SerialPort.BaseStream.ReadTimeout` and most of the time ignores `CancellationToken` instance. If you use `Task.Delay()` alongside `Task.WhenAny()` with `SerialPort.BaseStream.ReadAsync()` there is read task and byte buffer left to hang forever. So it doesn't seem like a good option. 

Another options is to use `SerialPort.DataRecieved` event. But in this case you basically probaby need to organize state machine so the event handler behaves differently depending on which command is currently in process.

The third option is to use synchronous methods and offload the work  to a Thread pool thread which seems the best of three:
```
static async Task OnClick(object sender, EventArgs args, CancellationToken cancellationToken)
{
    var tcs = new TaskCompletionSource<bool>();
    cancellationToken.Register(() => tcs.TrySetResult(true));

    var commandTask = Task.Run(() => {
        // Organize critical sections around logical serial port operations somehow.
        lock (_syncRoot) 
        {
            return SomeCommand();
        }
    });
    // Or use Task.Delay() instead of tcs.Task if no cancellation token.
    if (tcs.Task == await Task.WhenAny(tcs.Task, commandTask))
    {
        // Timeout;
        return;
    }
    var response = await commandTask;
    // Process response.
}

static byte[] SomeCommand()
{
    // Assume serial port timeouts are set.
    _serialPort.Write(...);

    int offset = 0;
    int count = ...; // Expected response length.
    byte[] buffer = new byte[count];
    while (count > 0)
    {
        var readCount = _serialPort.Read(buffer, offset, count);
        if (readCount > 0)
        {
            offset += readCount;
            count -= readCount;
        }
    }
    return buffer;
}
```

**Comment by LeonidVasilyev** [comment]
Thank you for the response.
> for reading/writing I personally prefer running a background thread and use synchronous APIs (especially reading - writing should be fine and intuitive asynchronously as well)

Is it correct that if you want to set timeout and exclude possibility that write operation will hang forever (possibly because not-ready-to-communicate pin is set by the device) holding byte arrays and tasks you still need to use synchronous API?

**Comment by krwq** [comment]
@LeonidVasilyev if you want to not hang then you'll need to pass CancellationToken which expires after desired time:

```
using (var cts = new CancellationTokenSource(1000))
{
   // pass cts.Token to async APIs
}
```

**Comment by LeonidVasilyev** [comment]
It seems like `SerialStream` ignores it as `NetworkStream` do. See [Document that CancellationToken in Stream.ReadAsync() is advisory ](https://github.com/dotnet/corefx/issues/25154) related issue. But the difference is that I am not aware of a way to cancel a `SerialPort` operation after timeout using some other `SerialPort` or `SerialStream` method.

**Comment by krwq** [comment]
@LeonidVasilyev could you share `dotnet --info`, csproj and sample code?

**Comment by IGx89** [comment]
We're banging our heads against this too, in our case with a SerialStream. We've tried applying https://devblogs.microsoft.com/pfxteam/how-do-i-cancel-non-cancelable-async-operations/ to it but that's deficient too.

**Comment by krwq** [comment]
@IGx89 which OS are you using? Could you share `dotnet --info`  output on the platform you experience issues?

**Comment by IGx89** [comment]
In our case it's both Windows 7 and Windows 10, running .NET Framework 4.7.2. But it looks like Microsoft hasn't changed the API/code/recommendations for .NET Core so the same deficiencies would be there too.

**Comment by krwq** [comment]
@IGx89 cancellation with SerialStream should work correctly on Linux but not sure about Windows at the moment (I think it should be fixable) - if full framework doesn't work as expected at this point we will likely not fix it.

**Comment by IGx89** [comment]
Makes sense, thanks for looking into it!

**Comment by mikkleini** [comment]
Moving to Linux is the answer? This is not encouraging at all.
The async API is there and it's not functioning (neither timeout or cancellation) so you got to do something about it.

**Comment by krwq** [comment]
@mikkleini as far as I understand the timeouts should work (SerialPort.Read/WriteTimeout) - I'd expect at minimum scenarios covered by MSDN samples to work. The cancellation is something we'd like to improve but we currently have many issues in different areas and SerialPort is not high on our priority list at the moment. Linux implementation was written from scratch so we have added the support and accounted for it when adding support for timeouts.

**Comment by mikkleini** [comment]
Yes, SerialPort ReadTimeout works when used with blocking SerialPort.Read() function. But that and BaseStream.ReadTimeout do not have effect on BaseStream.ReadAsync(). I tried using CancellationTokenSource with CancelAfter and that didn't work either (on Windows).

Interesting is that I was successfully using blocking read with threads and signals and then I thought I'll modernize it with TPL and the first thing after refactoring I saw was stuck ReadAsync() function. It's very rare to see a .NET FW/Core issue so obviously it sticks out. Since I have the working old code, it's not that critical, but I would prefer to see that the issue is acknowledged and fix is planned properly. Right now from this "question" and other similar topic #25154 it feels like it's not treated as issue at all.

**Comment by krwq** [comment]
@mikkleini this is an issue but it's not treated with any priority because there are currently ways to achieve sort of cancellation (as you mentioned above - they are not perfect but possible). Note that the Windows version of this code has existed in full framework for a very long time and we never got many complains about it which puts it much lower than other issues in other areas we have. This is something I'd personally like to get fixed but cannot pick up at the moment. If you have some time to make the improvements we would be happy to take the PRs but note that we will still need to do bunch of testing on different pieces of hardware to ensure there is no regressions since those are not currently tested in the CI.

**Comment by mikkleini** [comment]
I took a look into Unix SerialStream and Windows SerialStream + Stream implementations to figure out the issue. I was surprised to find IOLoop and Thread.Sleep() in SerialStream.Unix and tasks in Windows C# streams implementation. I thought the asynchronous stuff goes deeper, maybe even to OS level, but unless I miss something important, under the hood the job is still done in classical waiting threads method. That means there's probably no significant performance increase in using asynchronous interface versus using own thread with blocking functions. So even this piece of code could provide the "async" interface but with working read timeout:

```
public static class SerialPortExtension
{
    public static Task<int> ReadAsync(this SerialPort port, byte[] buffer, int offset, int count)
    {
        return Task.Run(() =>
        {
            return port.Read(buffer, offset, count);
        });
    }
}
```

I don't have time to do deep performance tests now, but @krwq could you please comment - does it make sense?

**Comment by krwq** [comment]
@mikkleini it can go to OS level, it really depends, on Unix it could partially be done by OS but there would be not much benefit since we would still be required to poll and wait - there are some improvements we could do: i.e. one loop per all ports but for SerialPort specifically it was not worth much effort since usually there is no more than few ports per machine anyway (as opposed to sockets or files)
